File: src/lib/agents/intentAgent.ts
import { SearchCriteria } from '../types';

/**
 * Intent Agent
 * Converts user input (form or natural language) into structured search criteria
 */
export class IntentAgent {
  static async processInput(input: {
    from: string;
    to: string;
    date: string;
    class: string;
    quota: string;
    berthPreference: string;
    tatkalTime?: string;
    passengerCount: number;
  }): Promise<SearchCriteria> {
    // Validate and normalize input
    const criteria: SearchCriteria = {
      from: input.from.toUpperCase(),
      to: input.to.toUpperCase(),
      date: input.date,
      class: input.class || 'AC_3',
      quota: input.quota || 'GENERAL',
      berthPreference: input.berthPreference || 'NO_PREFERENCE',
      tatkalTime: input.tatkalTime,
      passengerCount: Math.max(1, Math.min(6, input.passengerCount || 1)),
    };

    // Add logging
    console.log('[Intent Agent] Processing criteria:', criteria);

    return criteria;
  }

  static parseNaturalLanguage(text: string): Partial<SearchCriteria> {
    // Simple NLP parsing for natural language input
    const result: Partial<SearchCriteria> = {};

    // Extract train details from text
    const dateMatch = text.match(/(\d{1,2}[/-]\d{1,2}[/-]\d{4})/);
    if (dateMatch) result.date = dateMatch[0];

    const classMatch = text.match(/(sleeper|ac2|ac3)/i);
    if (classMatch) result.class = classMatch[1].toUpperCase();

    const berthMatch = text.match(/(lower|middle|upper|side)\s*(?:berth|bunk)?/i);
    if (berthMatch) result.berthPreference = berthMatch[1].toUpperCase();

    return result;
  }
}

File: backend/app/agents/intent_agent.py
"""
Intent Agent - Parses and validates user search criteria
"""
from pydantic import BaseModel
from typing import Optional

class SearchCriteria(BaseModel):
    from_city: str
    to_city: str
    date: str
    train_class: str = "ac3"
    quota: str = "GENERAL"
    berth_preference: str = "NO_PREFERENCE"
    tatkal_time: Optional[str] = None
    passenger_count: int = 1

class IntentAgent:
    @staticmethod
    def process_input(criteria: SearchCriteria) -> dict:
        """
        Parse and normalize user input
        """
        print(f"[Intent Agent] Processing: {criteria.from_city} → {criteria.to_city}")
        
        # Normalize input
        normalized = {
            "from_city": criteria.from_city.upper(),
            "to_city": criteria.to_city.upper(),
            "date": criteria.date,
            "train_class": criteria.train_class.lower() or "ac3",
            "quota": criteria.quota.upper() or "GENERAL",
            "berth_preference": criteria.berth_preference.upper() or "NO_PREFERENCE",
            "tatkal_time": criteria.tatkal_time or "08:00",
            "passenger_count": max(1, min(6, criteria.passenger_count)),
        }
        
        # Validate
        if not normalized["from_city"] or not normalized["to_city"]:
            raise ValueError("From and To cities are required")
        
        if normalized["from_city"] == normalized["to_city"]:
            raise ValueError("From and To cities cannot be the same")
        
        print(f"✓ [Intent Agent] Input validated and normalized")
        
        return normalized
    
    @staticmethod
    def parse_natural_language(text: str) -> dict:
        """
        Parse natural language queries (optional feature)
        """
        import re
        
        result = {}
        
        # Extract cities
        if "from" in text.lower() and "to" in text.lower():
            parts = text.lower().split("from")
            if len(parts) > 1:
                city_part = parts[1].split("to")
                if len(city_part) > 1:
                    result["from_city"] = city_part[0].strip()
                    result["to_city"] = city_part[1].split()[0].strip()
        
        # Extract class
        if "ac3" in text.lower():
            result["train_class"] = "ac3"
        elif "ac2" in text.lower():
            result["train_class"] = "ac2"
        elif "sleeper" in text.lower():
            result["train_class"] = "sleeper"
        
        # Extract berth
        if "lower" in text.lower():
            result["berth_preference"] = "LOWER"
        elif "middle" in text.lower():
            result["berth_preference"] = "MIDDLE"
        elif "upper" in text.lower():
            result["berth_preference"] = "UPPER"
        
        return result
